### PR TITLE
Datasets and loaders

### PR DESCRIPTION
**Pull request by JamesOwers**
First draft here. Rerun `make_dataset.py` and you'll get new files `{train,valid,test}_cmd_corpus.csv`. These files are read by the dataset in `mdtk.pytorch_datasets` to produce instances of a given index. For example:

```
from mdtk.pytorch_datasets import CommandDataset, CommandVocab
vocab = CommandVocab()
train_dataset = CommandDataset('./acme/train_cmd_corpus.csv', vocab, 128)
... Loading Dataset: 18005it [00:00, 144793.12it/s]
```
Each data item is a dictionary containing the degraded commands as integers (in a list), likewise for the clean commands, and the degradation label (I'm assuming that deg_label 0 will always mean 'not degraded', so we don't need a is_degraded flag).
```
train_dataset[0]
... {'deg_commands': [3, 49, 50, 1, 177, 1, 47, 64, 69, 1, 71, 1, 175, 1, 178, 45, 1, 173, 192, 1, 197, 1, 40, 59, 62, 64, 69, 73, 1, 187, 1, 192, 197, 201, 1, 64, 1, 71, 1, 168, 1, 47, 1, 192, 1, 199, 1, 175, 64, 69, 1, 40, 1, 192, 1, 168, 197, 1, 50, 64, 68, 68, 71, 76, 1, 178, 1, 199, 49, 1, 199, 1, 190, 196, 204, 1, 192, 1, 177, 196, 62, 1, 47, 64, 66, 69, 69, 74, 1, 175, 1, 194, 45, 1, 190, 197, 202, 1, 173, 1, 192, 197, 1, 40, 59, 64, 68, 68, 71, 76, 1, 187, 1, 168, 64, 1, 47, 1, 192, 1, 175, 1, 40, 64, 1, 168, 1, 192],
'clean_commands': [3, 49, 50, 1, 177, 1, 47, 64, 69, 1, 175, 1, 178, 45, 1, 173, 192, 1, 197, 1, 40, 59, 62, 64, 69, 73, 1, 187, 1, 192, 197, 201, 1, 64, 1, 71, 1, 168, 1, 47, 1, 192, 1, 199, 1, 175, 64, 69, 1, 40, 1, 192, 1, 168, 197, 1, 50, 64, 68, 68, 71, 76, 1, 178, 1, 199, 49, 1, 190, 196, 204, 1, 192, 1, 177, 196, 62, 1, 47, 64, 66, 69, 69, 74, 1, 175, 1, 194, 45, 1, 190, 197, 202, 1, 173, 1, 192, 197, 1, 40, 59, 64, 68, 68, 71, 71, 76, 1, 187, 1, 168, 64, 1, 47, 1, 192, 1, 175, 1, 40, 64, 1, 168, 1, 192, 50, 68, 1],
'deg_label': 2}
```

The vocabulary can be used to get the tokens back:
```
[vocab.itos[idx] for idx in train_dataset[0]['deg_commands']]
... ['<sos>', 'o45', 'o46', '<unk>', 'f45', '<unk>', 'o43', 'o60', 'o65', '<unk>', 'o67', '<unk>', 'f43', '<unk>', 'f46', 'o41', '<unk>', 'f41', 'f60', '<unk>', 'f65', '<unk>', 'o36', 'o55', 'o58', 'o60', 'o65', 'o69', '<unk>', 'f55', '<unk>', 'f60', 'f65', 'f69', '<unk>', 'o60', '<unk>', 'o67', '<unk>', 'f36', '<unk>', 'o43', '<unk>', 'f60', '<unk>', 'f67', '<unk>', 'f43', 'o60', 'o65', '<unk>', 'o36', '<unk>', 'f60', '<unk>', 'f36', 'f65', '<unk>', 'o46', 'o60', 'o64', 'o64', 'o67', 'o72', '<unk>', 'f46', '<unk>', 'f67', 'o45', '<unk>', 'f67', '<unk>', 'f58', 'f64', 'f72', '<unk>', 'f60', '<unk>', 'f45', 'f64', 'o58', '<unk>', 'o43', 'o60', 'o62', 'o65', 'o65', 'o70', '<unk>', 'f43', '<unk>', 'f62', 'o41', '<unk>', 'f58', 'f65', 'f70', '<unk>', 'f41', '<unk>', 'f60', 'f65', '<unk>', 'o36', 'o55', 'o60', 'o64', 'o64', 'o67', 'o72', '<unk>', 'f55', '<unk>', 'f36', 'o60', '<unk>', 'o43', '<unk>', 'f60', '<unk>', 'f43', '<unk>', 'o36', 'o60', '<unk>', 'f36', '<unk>', 'f60']
```

As you can see....that's rather more `<unk>` than I would like...

We can do most of the tasks with this data. And you can use the transform to get torch versions of the data if required.
```
from mdtk.pytorch_datasets import transform_to_torchtensor
train_dataset = CommandDataset('./acme/train_cmd_corpus.csv', vocab, 128, transform=transform_to_torchtensor)
train_dataset[0]
{'deg_commands': tensor([  3,  49,  50,   1, 177,   1,  47,  64,  69,   1,  71,   1, 175,   1,
        178,  45,   1, 173, 192,   1, 197,   1,  40,  59,  62,  64,  69,  73,
          1, 187,   1, 192, 197, 201,   1,  64,   1,  71,   1, 168,   1,  47,
          1, 192,   1, 199,   1, 175,  64,  69,   1,  40,   1, 192,   1, 168,
        197,   1,  50,  64,  68,  68,  71,  76,   1, 178,   1, 199,  49,   1,
        199,   1, 190, 196, 204,   1, 192,   1, 177, 196,  62,   1,  47,  64,
         66,  69,  69,  74,   1, 175,   1, 194,  45,   1, 190, 197, 202,   1,
        173,   1, 192, 197,   1,  40,  59,  64,  68,  68,  71,  76,   1, 187,
          1, 168,  64,   1,  47,   1, 192,   1, 175,   1,  40,  64,   1, 168,
          1, 192]), 'clean_commands': tensor([  3,  49,  50,   1, 177,   1,  47,  64,  69,   1, 175,   1, 178,  45,
          1, 173, 192,   1, 197,   1,  40,  59,  62,  64,  69,  73,   1, 187,
          1, 192, 197, 201,   1,  64,   1,  71,   1, 168,   1,  47,   1, 192,
          1, 199,   1, 175,  64,  69,   1,  40,   1, 192,   1, 168, 197,   1,
         50,  64,  68,  68,  71,  76,   1, 178,   1, 199,  49,   1, 190, 196,
        204,   1, 192,   1, 177, 196,  62,   1,  47,  64,  66,  69,  69,  74,
          1, 175,   1, 194,  45,   1, 190, 197, 202,   1, 173,   1, 192, 197,
          1,  40,  59,  64,  68,  68,  71,  71,  76,   1, 187,   1, 168,  64,
          1,  47,   1, 192,   1, 175,   1,  40,  64,   1, 168,   1, 192,  50,
         68,   1]), 'deg_label': tensor(2)}
```

With this transformed data, the dataset can happily be used with the vanilla pytorch dataloader for batching:
```
from torch.utils.data import DataLoader
train_dataloader = DataLoader(train_dataset, batch_size=4,
                                                     shuffle=True, num_workers=4)
for batch_nr, batch in enumerate(train_dataloader):
    print(f'batch nr {batch_nr}')
    print(f'The batch is a {type(batch)}')
    print('But its values are pytorch tensors')
    labels = batch['deg_label']
    in_tokens = batch['deg_commands']
    print(labels.shape, in_tokens.shape)
    if batch_nr > 3:
        break
```

```
batch nr 0
The batch is a <class 'dict'>
But its values are pytorch tensors
torch.Size([4]) torch.Size([4, 128])
batch nr 1
The batch is a <class 'dict'>
But its values are pytorch tensors
torch.Size([4]) torch.Size([4, 128])
batch nr 2
The batch is a <class 'dict'>
But its values are pytorch tensors
torch.Size([4]) torch.Size([4, 128])
batch nr 3
The batch is a <class 'dict'>
But its values are pytorch tensors
torch.Size([4]) torch.Size([4, 128])
batch nr 4
The batch is a <class 'dict'>
But its values are pytorch tensors
torch.Size([4]) torch.Size([4, 128])
```

Finally, you can use the [nn.Embedding](https://pytorch.org/docs/stable/nn.html#torch.nn.Embedding) layer as the first layer to the net to get from these integers to one hots (or learned vectors...). For an example, see how they do token embedding with BERT [here](https://github.com/codertimo/BERT-pytorch/blob/master/bert_pytorch/model/embedding/bert.py) (it calls out the [here](https://github.com/codertimo/BERT-pytorch/blob/master/bert_pytorch/model/embedding/token.py))

TODO:
* Implement rounding to nearest time_increment in `df_to_command_str`
* handle cases where there is a longer pause than `max_time_shift` in `df_to_command_str` (just output multiple time shift commands in a row
* Overlapping notes mess up the conversion back from commands to df (see issue #20 )

### CODE REVIEW

File: make_dataset.py [hunk 1]
```diff
@@ -483,25 +500,60 @@ def parse_args(args_input=None):
...
+                               names=['onset', 'track', 'pitch', 'dur'])
+        clean_cmd_str = df_to_command_str(clean_df)
+        deg_num = row.degradation_id
+        split = row.split
+        fh = fh_dict[split]
+        fh.write(f'{alt_cmd_str},{clean_cmd_str},{deg_num}\n')
+        meta_df.loc[idx, 'corpus_path'] = fh.name
+        meta_df.loc[idx, 'corpus_line_nr'] = line_counts[split]
+        line_counts[split] += 1
+    meta_df.to_csv(os.path.join(ARGS.output_dir, 'metadata.csv'))
```
Comment (apmcleod) [inline make_dataset.py]: I think this should be optional with a --command flag, since it more than doubles runtime, and not everyone will want to use the command-based one-hots that we use.

We should make a function somewhere, like meta_csv_to_cmd_csvs() and just call that from the Dataset if it hasn't yet been run (or call that func from here if --command is True)

File: make_dataset.py [hunk 2]
```diff
@@ -13,6 +13,7 @@
 
 from mdtk import degradations, downloaders, data_structures, midi
 from mdtk.filesystem_utils import make_directory, copy_file
+from mdtk.pytorch_datasets import df_to_command_str
```
Comment (apmcleod) [inline make_dataset.py]: This import requires pytorch to be installed. See #54

File: mdtk/pytorch_datasets.py [hunk 3]
```diff
@@ -1 +1,283 @@
...
+    commands = pd.concat((note_on, note_off)).sort_values(
+                   ['onset', 'cmd_type', 'pitch'],
+                   ascending=[True, True, True])
+    
+    command_list = []
+    current_onset = commands.onset.iloc[0]
+    for idx, row in commands.iterrows():
+        time_shift = row.onset - current_onset
+        if time_shift != 0:
+            command_list += [f't{time_shift}']
```
Comment (apmcleod) [inline mdtk/pytorch_datasets.py]: This time shift is in ms, not frames, which causes all of the <unk>s.

File: make_dataset.py [hunk 4]
```diff
@@ -483,25 +500,60 @@ def parse_args(args_input=None):
...
     meta_file.close()
+    
+    fh_dict = {
+        split: open(
+            os.path.join(ARGS.output_dir, f'{split}_cmd_corpus.csv'
+        ), 'w') for split in ['train', 'valid', 'test']
+    }
+    line_counts = {
+        split: 0 for split in ['train', 'valid', 'test']
+    }
```
Comment (apmcleod) [inline make_dataset.py]: This is already in current_splits, but not explicitly as a dict, and this doesn't add much time anyways I'm sure.

File: make_dataset.py [hunk 5]
```diff
@@ -483,25 +500,60 @@ def parse_args(args_input=None):
...
+    meta_df = pd.read_csv(os.path.join('acme', 'metadata.csv'))
+    for idx, row in tqdm(meta_df.iterrows(), total=meta_df.shape[0],
+                         desc='Creating command corpus'):
+        alt_df = pd.read_csv(os.path.join('acme', row.altered_csv_path),
+                             header=None,
+                             names=['onset', 'track', 'pitch', 'dur'])
+        alt_cmd_str = df_to_command_str(alt_df)
+        clean_df = pd.read_csv(os.path.join('acme', row.clean_csv_path),
+                               header=None,
+                               names=['onset', 'track', 'pitch', 'dur'])
```
Comment (apmcleod) [inline make_dataset.py]: Does data_structures.read_note_csv not work for this? If not, we need a function that does, I think.

File: make_dataset.py (hunk 1 above)
Comment (apmcleod) [reply make_dataset.py]: We should also move this into the degradation loop to avoid the writing/reading the csvs multiple times, which probably causes a lot of the slowdown.

File: mdtk/pytorch_datasets.py [hunk 6]
```diff
@@ -1 +1,283 @@
...
+        changed.index.names = ['id', 'col']
+        difference_locations = np.where(diff_mask)
+        changed_from = df1.values[difference_locations]
+        changed_to = df2.values[difference_locations]
+        return pd.DataFrame({'from': changed_from, 'to': changed_to},
+                            index=changed.index)
+
+# TODO: later can auto detect vocab from corpus if necessary
+#       I'm doing things this way just for ability to change things
+#       later with ease
```
Comment (apmcleod) [inline mdtk/pytorch_datasets.py]: I'm not sure it would be reasonably possible to detect the vocab later. We can read and find all of the used vocab symbols (as ints), but it will be much more complicated to figure out exactly min/max pitch, and time_increment.

We should write the params for this later.

File: make_dataset.py (hunk 1 above)
Comment (apmcleod) [reply make_dataset.py]: I added --command arg and moved the func elsewhere to allow post-processing (and it works)

File: make_dataset.py (hunk 2 above)
Comment (apmcleod) [reply make_dataset.py]: You mention this in a TODO as well.

File: mdtk/pytorch_datasets.py (hunk 3 above)
Comment (apmcleod) [reply mdtk/pytorch_datasets.py]: Ok, now that I looked more closely. This time shift is not quantized, which leads to the unks. I fixed in a future commit.

File: mdtk/pytorch_datasets.py (hunk 6 above)
Comment (apmcleod) [reply mdtk/pytorch_datasets.py]: Oh, nevermind! I see the csv contain strings, and the loader converts to ints.

**Comment by apmcleod** [comment]
lgtm. Made a few changes.

Regarding your TODOs:

-Implement rounding to nearest time_increment in df_to_command_str: **DONE**
-handle cases where there is a longer pause than max_time_shift in df_to_command_str (just output multiple time shift commands in a row: **DONE**
-Overlapping notes mess up the conversion back from commands to df (see issue #20 ): Complex, also see issues #46 and #8.

I'll let you check over this and complete the pull whenever you're ready. In the meantime I'll branch off of this point and continue working.

File: mdtk/pytorch_datasets.py [hunk 7]
```diff
@@ -1 +1,327 @@
...
+    def get_corpus_line(self, item):
+        if self.in_memory:
+            deg_cmd, clean_cmd, deg_num = self.lines[item]
+            return deg_cmd, clean_cmd, deg_num
+        else:
+            line = self.file.__next__()
+            if line is None:
+                self.file.close()
+                self.file = open(self.corpus_path, "r", encoding=self.encoding)
+                line = self.file.__next__()
```
Comment (apmcleod) [inline mdtk/pytorch_datasets.py]: This won't work with shuffling, correct?

File: mdtk/pytorch_datasets.py [hunk 8]
```diff
@@ -1 +1,327 @@
...
+        transform: func
+            The output from __get_item__ is a dictionary of numpy arrays.
+            The function transform is applied to the dictionary before it is
+            returned so, for example, it can be used to convert all data to
+            torch tensors.
+    
+        Returns
+        -------
+        df : pd.DataFrame
+            The pandas DataFrame representing the note data
```
Comment (apmcleod) [inline mdtk/pytorch_datasets.py]: I don't think it returns anything.

File: mdtk/pytorch_datasets.py (hunk 7 above)
Comment (apmcleod) [reply mdtk/pytorch_datasets.py]: Ah, you mentioned that in the docstring!

File: make_dataset.py (hunk 1 above)
Comment (JamesOwers) [reply make_dataset.py]: Sure, though I slightly preference the default doing it anyway but we can tussle over that later.

File: make_dataset.py (hunk 2 above)
Comment (JamesOwers) [reply make_dataset.py]: yeah, I think we agreed it was ok to have this as a dependency? If not, we should move those functions out of pytorch_datasets into another module (i'd prefer this anyway)

File: mdtk/pytorch_datasets.py (hunk 3 above)
Comment (JamesOwers) [reply mdtk/pytorch_datasets.py]: yeah this is where rounding is required

File: make_dataset.py (hunk 5 above)
Comment (JamesOwers) [reply make_dataset.py]: not sure I completely follow

File: mdtk/pytorch_datasets.py (hunk 6 above)
Comment (JamesOwers) [reply mdtk/pytorch_datasets.py]: Yeah, this is what the initial vocab does - just defines all possible strings expected in commands. However, this is not very efficient - there may be some strings not present in the data, and if we just want to read the data and find all possible tokens (no rounding of ms to nearest 40ms or whatever), then you can do that relatively easily. That's all I mean by this comment.

File: mdtk/pytorch_datasets.py (hunk 7 above)
Comment (JamesOwers) [reply mdtk/pytorch_datasets.py]: haha! Yes! It's the functionality i've adapted from https://github.com/codertimo/BERT-pytorch. It's not ideal but at least something is in place that we can change later (I'm sure there must be a way of reading a specific line from a file (I think `sed` can do it for instance)

File: mdtk/pytorch_datasets.py (hunk 8 above)
Comment (JamesOwers) [reply mdtk/pytorch_datasets.py]: Haha, yes, sorry I copied a docstring above and forgot to remove this. I'll remove.

File: make_dataset.py (hunk 1 above)
Comment (apmcleod) [reply make_dataset.py]: I actually have it as the default now (on models branch)

File: make_dataset.py (hunk 2 above)
Comment (apmcleod) [reply make_dataset.py]: Yeah, the dependency is ok, but in case people can't, and are trying to run this script without pytorch somehow I mean. Moving it into another module is perfect.

**Comment by JamesOwers** [comment]
Phew, that was a big one. I'm loving this process man, thanks.

I think the most important thing to sort is the overlapping pitches problem. The pipeline for dataset creation and giving an example model should determine what we should do with the `--command` flag. I think I'm agreed with you now really, I've no idea how people will want to use this data!

File: make_dataset.py (hunk 5 above)
Comment (apmcleod) [reply make_dataset.py]: There's already a function data_structures.read_note_csv that I thought we had written to do this.

File: mdtk/pytorch_datasets.py (hunk 6 above)
Comment (apmcleod) [reply mdtk/pytorch_datasets.py]: The rounding I implemented actually now pre-quantizes everything before writing to the csv, so converting back out without quantization is impossible, if we keep it. But I made the convert metadata to cmd csv its own function so people can easily create unquantized vocab that way.
